File: python/m5/l2/hw_answers/agent/graph.py
from langchain_core.messages import AIMessage
from langchain_core.runnables import RunnableConfig
from langgraph.graph import END, START, MessagesState, StateGraph


def chat(state: MessagesState, config: RunnableConfig) -> dict:
    user = config["configurable"].get("langgraph_auth_user", {})
    user_info = ", ".join(f"{k}={v}" for k, v in user.items())
    last = state["messages"][-1].content
    return {"messages": [AIMessage(content=f"[{user_info}] {last}")]}


graph = StateGraph(MessagesState)
graph.add_node("chat", chat)
graph.add_edge(START, "chat")
graph.add_edge("chat", END)
graph = graph.compile()

File: typescript/m5/l2/hw_answers/agent/graph.ts
import { END, START, MessagesAnnotation, StateGraph, type LangGraphRunnableConfig } from "@langchain/langgraph";
import { AIMessage } from "@langchain/core/messages";

// Fields the TS SDK injects automatically — skip them so the output matches
// what the auth handler explicitly returned (identity, name, role).
const SKIP_KEYS = new Set(["permissions", "is_authenticated", "display_name"]);

function chat(
  state: typeof MessagesAnnotation.State,
  config: LangGraphRunnableConfig,
): typeof MessagesAnnotation.Update {
  const user = (config.configurable?.langgraph_auth_user ?? {}) as Record<string, unknown>;
  const userInfo = Object.entries(user)
    .filter(([k]) => !SKIP_KEYS.has(k))
    .map(([k, v]) => `${k}=${v}`)
    .join(", ");
  const last = state.messages[state.messages.length - 1];
  const content = typeof last.content === "string" ? last.content : "";
  return { messages: [new AIMessage(`[${userInfo}] ${content}`)] };
}

export const graph = new StateGraph(MessagesAnnotation)
  .addNode("chat", chat)
  .addEdge(START, "chat")
  .addEdge("chat", END)
  .compile();
